From e33c1b80bfe81dfd90c7a1d68b28ac7fa861f737 Mon Sep 17 00:00:00 2001
From: "Russ W. Knize" <Russ.Knize@motorola.com>
Date: Thu, 4 Dec 2014 16:42:23 -0600
Subject: [PATCH 473/959] IKSWM-11228: mmc: refresh card identification for MMC
 cards

One or more CID entries may change on a card whose firmware has been
reflashed.  If the raw CID buffer has been zeroed-out, just assume
that the CID, CSD, and EXT_CSD need to be re-read and re-parsed.

Change-Id: I377ae49e69444cbdc95bb558ce24071fd0402853
Signed-off-by: Russ Knize <Russ.Knize@motorola.com>
Reviewed-on: http://gerrit.mot.com/764643
SME-Granted: SME Approvals Granted
SLTApproved: Slta Waiver <sltawvr@motorola.com>
Tested-by: Jira Key <jirakey@motorola.com>
Reviewed-by: Igor Kovalenko <igork@motorola.com>
Submit-Approved: Jira Key <jirakey@motorola.com>
Reviewed-on: http://gerrit.mot.com/799510
Reviewed-by: Russell Knize <rknize@motorola.com>
---
 drivers/mmc/core/mmc.c | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/drivers/mmc/core/mmc.c b/drivers/mmc/core/mmc.c
index 3d6e63c5a20..c524e3184fd 100644
--- a/drivers/mmc/core/mmc.c
+++ b/drivers/mmc/core/mmc.c
@@ -1520,6 +1520,7 @@ static int mmc_init_card(struct mmc_host *host, u32 ocr,
 	u32 cid[4];
 	u32 rocr;
 	u8 *ext_csd = NULL;
+	bool scan = (oldcard == NULL);
 
 	BUG_ON(!host);
 	WARN_ON(!host->claimed);
@@ -1572,6 +1573,15 @@ reinit:
 	}
 
 	if (oldcard) {
+		if (oldcard->raw_cid[0] == 0 && oldcard->raw_cid[1] == 0 &&
+		    oldcard->raw_cid[2] == 0 && oldcard->raw_cid[3] == 0) {
+			scan = true;
+			pr_info("%s: updating card identification\n", mmc_hostname(host));
+			memcpy(oldcard->raw_cid, cid, sizeof(oldcard->raw_cid));
+			err = mmc_decode_cid(oldcard);
+			if (err)
+				goto err;
+		}
 		if (memcmp(cid, oldcard->raw_cid, sizeof(cid)) != 0) {
 			err = -ENOENT;
 			pr_err("%s: %s: CID memcmp failed %d\n",
@@ -1613,7 +1623,7 @@ reinit:
 		mmc_set_bus_mode(host, MMC_BUSMODE_PUSHPULL);
 	}
 
-	if (!oldcard) {
+	if (scan) {
 		/*
 		 * Fetch CSD from card.
 		 */
@@ -1650,7 +1660,7 @@ reinit:
 		}
 	}
 
-	if (!oldcard) {
+	if (scan) {
 		/*
 		 * Fetch and process extended CSD.
 		 */
-- 
2.11.0

